mmdet/models/detectors/deformable_detr.py

```diff
@@ -1,12 +1,598 @@
+from mmengine.model import ModuleList, xavier_init
+from torch import Tensor, nn
+from torch.nn.init import normal_
+
 from mmdet.registry import MODELS
-from .detr import DETR
+from mmdet.structures import OptSampleList, SampleList
+from mmdet.utils import OptConfigType
+from ..layers import (DetrTransformerDecoder, DetrTransformerDecoderLayer,
+                      DetrTransformerEncoder, DetrTransformerEncoderLayer,
+                      SinePositionalEncoding, inverse_sigmoid)
+from .base_detr import TransformerDetector
 
 
 @MODELS.register_module()
-class DeformableDETR(DETR):
+class DeformableDETR(TransformerDetector):
     r"""Implementation of `Deformable DETR: Deformable Transformers for
     End-to-End Object Detection <https://arxiv.org/abs/2010.04159>`_"""
 
-    def __init__(self, *args, **kwargs) -> None:
-        super(DETR, self).__init__(*args, **kwargs)
+    def __init__(self,
+                 *args,
+                 decoder_cfg: OptConfigType = None,
+                 bbox_head: OptConfigType = None,
+                 with_box_refine: bool = False,
+                 as_two_stage: bool = False,
+                 num_feature_levels: int = 4,
+                 **kwargs) -> None:
+        self.with_box_refine = with_box_refine
+        self.as_two_stage = as_two_stage
+        self.num_feature_levels = num_feature_levels
+
+        if "with_box_refine" in bbox_head:
+            assert bbox_head.with_box_refine == with_box_refine
+        else:
+            bbox_head.with_box_refine = with_box_refine
+        if "as_two_stage" in bbox_head:
+            assert bbox_head.as_two_stage == as_two_stage
+        else:
+            bbox_head.as_two_stage = as_two_stage
+        if "num_decoder_layers" in bbox_head:
+            assert bbox_head.num_decoder_layers == decoder_cfg.num_layers
+        else:
+            bbox_head.num_decoder_layers = decoder_cfg.num_layers
+
+        super().__init__(
+            *args, decoder_cfg=decoder_cfg, bbox_head=bbox_head, **kwargs)
+
+    def _init_layers(self) -> None:
+        self.positional_encoding = SinePositionalEncoding(
+            **self.positional_encoding_cfg)
+        self.encoder = DeformableDetrTransformerEncoder(**self.encoder_cfg)
+        self.decoder = DeformableDetrTransformerDecoder(**self.decoder_cfg)
+        self.embed_dims = self.encoder.embed_dims
+        if not self.as_two_stage:
+            self.query_embedding = nn.Embedding(self.num_query,
+                                                self.embed_dims * 2)
+
+        num_feats = self.positional_encoding.num_feats
+        assert num_feats * 2 == self.embed_dims, \
+            f'embed_dims should be exactly 2 times of num_feats. ' \
+            f'Found {self.embed_dims} and {num_feats}.'
+
+        self.level_embeds = nn.Parameter(
+            torch.Tensor(self.num_feature_levels, self.embed_dims))
+
+        if self.as_two_stage:
+            self.enc_output = nn.Linear(self.embed_dims, self.embed_dims)
```

Comment:
enc_output_fc

Reply:
Have you fixed it？

Reply:
> Have you fixed it？

I missed committing this variable and the enc_output has now been renamed with memory_trans_fc and the enc_output_norm has been renamed with memory_trans_norm.